src/validation/validation.cpp

```diff
@@ -0,0 +1,3126 @@
+
+#include "validation.h"
+
+#include "consensus/merkle.h"
+#include "consensus/tx_verify.h"
+
+#include "blockstorage/blockstorage.h"
+#include "blockstorage/sequential_files.h"
+#include "requestManager.h"
+#include "sync.h"
+#include "ui_interface.h"
+#include "validationinterface.h"
+
+#include "checkpoints.h"
+#include "expedited.h"
+#include "init.h"
+
+#include "connmgr.h"
+#include "dosman.h"
+
+#include "timedata.h"
+#include "txorphanpool.h"
+
+#include "txadmission.h"
+
+#include <boost/scope_exit.hpp>
+
+struct CBlockIndexWorkComparator
+{
+    bool operator()(CBlockIndex *pa, CBlockIndex *pb) const
+    {
+        // First sort by most total work, ...
+        if (pa->nChainWork > pb->nChainWork)
+            return false;
+        if (pa->nChainWork < pb->nChainWork)
+            return true;
+
+        // ... then by earliest time received, ...
+        if (pa->nSequenceId < pb->nSequenceId)
+            return false;
+        if (pa->nSequenceId > pb->nSequenceId)
+            return true;
+
+        // Use pointer address as tie breaker (should only happen with blocks
+        // loaded from disk, as those all have id 0).
+        if (pa < pb)
+            return false;
+        if (pa > pb)
+            return true;
+
+        // Identical blocks.
+        return false;
+    }
+};
+
+// bip135 begin
+// keep track of count over last 100
+struct UnknownForkData
+{
+    int UnknownForkSignalStrength{0};
+    bool UnknownForkSignalFirstDetected{false};
+    bool UnknownForkSignalLost{false};
+    bool UnknownForkSignalAt25Percent{false};
+    bool UnknownForkSignalAt50Percent{false};
+    bool UnknownForkSignalAt70Percent{false};
+    bool UnknownForkSignalAt90Percent{false};
+    bool UnknownForkSignalAt95Percent{false};
+};
+
+static UnknownForkData unknownFork[Consensus::MAX_VERSION_BITS_DEPLOYMENTS];
+// bip135 end
+
+/**
+ * The set of all CBlockIndex entries with BLOCK_VALID_TRANSACTIONS (for itself and all ancestors) and
+ * as good as our current tip or better. Entries may be failed, though, and pruning nodes may be
+ * missing the data for the block.
+ */
+std::set<CBlockIndex *, CBlockIndexWorkComparator> setBlockIndexCandidates GUARDED_BY(cs_main);
+
+// Last time the block tip was updated
+std::atomic<int64_t> nTimeBestReceived{0};
+
+CBlockIndex *pindexBestForkTip = nullptr;
+CBlockIndex *pindexBestForkBase = nullptr;
+
+extern CCriticalSection cs_recentRejects;
```

Comment:
not used

Reply:
pindexBestForkTip is used in CheckForkWarningConditions() defined on line 1331
pindexBestForkBase is used on line 2665

however recentRejects isnt used. i will fix that.